File: java/proxy/DummyThirdPartyApiService.java
package proxy;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.HashMap;
import java.util.Map;

public class DummyThirdPartyApiService implements DummyJsonApi {

    private static final String PRODUCTS_ENDPOINT = "https://dummyjson.com/products";
    private static final String RECIPES_ENDPOINT = "https://dummyjson.com/recipes";
    public DummyThirdPartyApiService() {
    }

    public String getAllProducts() {
        try {
            System.out.println("Fetching products...");
            Thread.sleep(2000);
            HttpRequest httpRequest = HttpRequest.newBuilder()
                    .uri(new URI(PRODUCTS_ENDPOINT))
                    .GET()
                    .build();
            HttpClient httpClient = HttpClient.newBuilder().build();
            return httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString()).body();
        } catch (URISyntaxException | IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public String getAllRecipes() {
        try {
            System.out.println("Fetching recipes...");
            Thread.sleep(2000);
            HttpRequest httpRequest = HttpRequest.newBuilder()
                    .uri(new URI(RECIPES_ENDPOINT))
                    .GET()
                    .build();
            HttpClient httpClient = HttpClient.newBuilder().build();
            return httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString()).body();
        } catch (URISyntaxException | IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: C#/Proxy/DummyThirdPartyApiService.cs
namespace Proxy;

/// <summary>
/// The <see cref="DummyThirdPartyApiService"/> contains some core business logic. Usually,
/// the classes are capable of doing some useful work which may also be very slow or
/// sensitive - e.g. correcting input data. A Proxy can solve these issues
/// without any changes to the <see cref="DummyThirdPartyApiService"/>'s code.
/// </summary>
public sealed class DummyThirdPartyApiService : IDummyJsonApi
{
    private static readonly HttpClient HttpClient = new();

    public string GetAllProducts()
    {
        try
        {
            Console.WriteLine("Fetching Products...");
            Thread.Sleep(2000);

            // Usage of Result is not recommended in production code
            // use async/await instead
            var response = HttpClient.GetAsync("https://dummyjson.com/products/").Result;
            var productsString =  response.Content.ReadAsStringAsync().Result;
            return productsString;
        }
        catch (Exception)
        {
            throw;
        }
    }

    public string GetAllRecipes()
    {
        try
        {
            Console.WriteLine("Fetching Recipes...");
            Thread.Sleep(2000);

            // Usage of Result is not recommended in production code
            // use async/await instead
            var response = HttpClient.GetAsync("https://dummyjson.com/recipes/").Result;
            var productsString = response.Content.ReadAsStringAsync().Result;
            return productsString;
        }
        catch (Exception)
        {
            throw;
        }
    }
}
